src-self-hosted/Module.zig

```diff
@@ -515,13 +528,47 @@ pub const Scope = struct {
         zir_module,
         /// .zig source code.
         file,
+        /// struct, enum or union, every .file contains one of these.
+        container,
         block,
         decl,
         gen_zir,
         local_val,
         local_ptr,
     };
 
+    pub const Container = struct {
+        pub const base_tag: Tag = .container;
+        base: Scope = Scope{ .tag = base_tag },
+
+        file_scope: *Scope.File,
+
+        /// Direct children of the file.
+        decls: ArrayListUnmanaged(*Decl),
```

Comment:
This can be a `std.AutoArrayHashMapUnmanaged`. It's roughly the same memory overhead as an ArrayListUnmanaged and it maintains insertion order. It will make the `removeDecl` logic simpler (and better big-O) below.

Edit: I see that you moved this code from elsewhere :+1: 
I was planning to do this change and overlooked it. Will do it upon merging this PR.